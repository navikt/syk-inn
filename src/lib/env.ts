import * as z from 'zod'

import { KeysOfUnion, raise } from './ts'

export type BundledEnv = z.infer<typeof BundledEnvSchema>
const BundledEnvSchema = z.object({
    runtimeEnv: z.union([
        z.literal('local'),
        z.literal('e2e'),
        z.literal('demo'),
        z.literal('prod-gcp'),
        z.literal('dev-gcp'),
    ]),
    NEXT_PUBLIC_BASE_PATH: z.string().nullish(),
    NEXT_PUBLIC_ASSET_PREFIX: z.string().nullish(),
    NEXT_PUBLIC_TELEMETRY_URL: z.string().nullish(),
    NEXT_PUBLIC_VERSION: z.string().nullish(),
})

/**
 * Next.js will bundle any environment variables that start with NEXT_PUBLIC_ into the
 * client bundle. These are available at any time, statically, in both server and browser.
 */
export const bundledEnv = BundledEnvSchema.parse({
    runtimeEnv: process.env.NEXT_PUBLIC_RUNTIME_ENV,
    NEXT_PUBLIC_BASE_PATH: process.env.NEXT_PUBLIC_BASE_PATH,
    NEXT_PUBLIC_ASSET_PREFIX: process.env.NEXT_PUBLIC_ASSET_PREFIX,
    NEXT_PUBLIC_TELEMETRY_URL: process.env.NEXT_PUBLIC_TELEMETRY_URL,
    NEXT_PUBLIC_VERSION: process.env.NEXT_PUBLIC_VERSION,
} satisfies Record<keyof BundledEnv, unknown>)

type ValkeyConfig = z.infer<typeof ValkeyConfigSchema>
const ValkeyConfigSchema = z.union([
    /**
     * Defines a union type for strongly typing Valkey configurations for local and production environments.
     * The local setup doesn't require authentication but does need the Docker image URL.
     */
    z.object({
        runtimeEnv: z.union([z.literal('dev-gcp'), z.literal('prod-gcp')]),
        username: z.string(),
        password: z.string(),
        tls: z.object({
            host: z.string(),
            port: z.coerce.number(),
        }),
    }),
    z.object({
        runtimeEnv: z.literal('local'),
        host: z.string(),
    }),
])

type ServerEnv = z.infer<typeof ServerEnvSchema>
const ServerEnvSchema = z.object({
    valkeyConfig: ValkeyConfigSchema.nullish(),
    useLocalSykInnApi: z.boolean().default(false),
    useLocalValkey: z.boolean().default(false),
})

/**
 * Pure server environment variables (i.e. not prefixed with NEXT_PUBLIC_) are only available
 * at runtime, and only on the server. Because these are strongly typed, they need to be accessed
 * lazily, otherwise the build would fail because of Next.js aggressive static optimizations.
 *
 * This can also be used in /api/internal/is_ready to verify that
 * the server is configured correctly before receiving any traffic.
 */
export function getServerEnv(): ServerEnv {
    const valkeyConfig =
        bundledEnv.runtimeEnv !== 'demo' && bundledEnv.runtimeEnv !== 'e2e'
            ? ({
                  runtimeEnv: process.env.NEXT_PUBLIC_RUNTIME_ENV,
                  username: process.env.VALKEY_USERNAME_SYK_INN,
                  password: process.env.VALKEY_PASSWORD_SYK_INN,
                  // Local
                  host: process.env.VALKEY_HOST_SYK_INN,
                  // Cloud
                  tls: {
                      host: process.env.VALKEY_HOST_SYK_INN,
                      port: process.env.VALKEY_PORT_SYK_INN,
                  },
              } satisfies Record<KeysOfUnion<ValkeyConfig>, unknown | undefined>)
            : undefined

    const parsedEnv = ServerEnvSchema.parse({
        valkeyConfig: valkeyConfig,
        useLocalSykInnApi: process.env.USE_LOCAL_SYK_INN_API === 'true',
        useLocalValkey: process.env.USE_LOCAL_VALKEY === 'true',
    } satisfies Record<keyof ServerEnv, unknown | undefined>)

    if (bundledEnv.runtimeEnv !== 'local' && parsedEnv.useLocalSykInnApi) {
        raise('USE_LOCAL_SYK_INN_API should only be set to true in local environment')
    }

    return parsedEnv
}

export const isE2E = bundledEnv.runtimeEnv === 'e2e'
export const isDemo = bundledEnv.runtimeEnv === 'demo'
export const isLocal = bundledEnv.runtimeEnv === 'local'
export const isCloud = bundledEnv.runtimeEnv === 'dev-gcp' || bundledEnv.runtimeEnv === 'prod-gcp'
