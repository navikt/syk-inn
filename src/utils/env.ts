import { z } from 'zod'

import { KeysOfUnion } from '@utils/ts'

export type BundledEnv = z.infer<typeof BundledEnvSchema>
const BundledEnvSchema = z.object({
    NEXT_PUBLIC_RUNTIME_ENV: z.union([
        z.literal('local'),
        z.literal('e2e'),
        z.literal('demo'),
        z.literal('prod-gcp'),
        z.literal('dev-gcp'),
    ]),
    NEXT_PUBLIC_BASE_PATH: z.string().nullish(),
    NEXT_PUBLIC_ASSET_PREFIX: z.string().nullish(),
})

/**
 * Next.js will bundle any environment variables that start with NEXT_PUBLIC_ into the
 * client bundle. These are available at any time, statically, in both server and browser.
 */
export const bundledEnv = BundledEnvSchema.parse({
    NEXT_PUBLIC_RUNTIME_ENV: process.env.NEXT_PUBLIC_RUNTIME_ENV,
    NEXT_PUBLIC_BASE_PATH: process.env.NEXT_PUBLIC_BASE_PATH,
    NEXT_PUBLIC_ASSET_PREFIX: process.env.NEXT_PUBLIC_ASSET_PREFIX,
})

type ValkeyConfig = z.infer<typeof ValkeyConfigSchema>
const ValkeyConfigSchema = z.union([
    /**
     * Defines a union type for strongly typing Valkey configurations for local and production environments.
     * The local setup doesn't require authentication but does need the Docker image URL.
     */
    z.object({
        runtimeEnv: z.union([z.literal('dev-gcp'), z.literal('prod-gcp')]),
        username: z.string(),
        password: z.string(),
        tls: z.object({
            host: z.string(),
            port: z.coerce.number(),
        }),
    }),
    z.object({
        runtimeEnv: z.literal('local'),
        host: z.string(),
    }),
])

type ServerEnv = z.infer<typeof ServerEnvSchema>
const ServerEnvSchema = z.object({
    helseIdWellKnown: z.string(),
    valkeyConfig: ValkeyConfigSchema.nullish(),
})

/**
 * Pure server environment variables (i.e. not prefixed with NEXT_PUBLIC_) are only available
 * at runtime, and only on the server. Because these are strongly typed, they need to be accessed
 * lazily, otherwise the build would fail because of Next.js aggressive static optimizations.
 *
 * This can also be used in /api/internal/is_ready to verify that
 * the server is configured correctly before receiving any traffic.
 */
export function getServerEnv(): ServerEnv {
    const valkeyConfig =
        bundledEnv.NEXT_PUBLIC_RUNTIME_ENV !== 'demo' && bundledEnv.NEXT_PUBLIC_RUNTIME_ENV !== 'e2e'
            ? ({
                  runtimeEnv: process.env.NEXT_PUBLIC_RUNTIME_ENV,
                  username: process.env.VALKEY_USERNAME_SYK_INN,
                  password: process.env.VALKEY_PASSWORD_SYK_INN,
                  // Local
                  host: process.env.VALKEY_HOST_SYK_INN,
                  // Cloud
                  tls: {
                      host: process.env.VALKEY_HOST_SYK_INN,
                      port: process.env.VALKEY_PORT_SYK_INN,
                  },
              } satisfies Record<KeysOfUnion<ValkeyConfig>, unknown | undefined>)
            : undefined

    return ServerEnvSchema.parse({
        valkeyConfig: valkeyConfig,
        helseIdWellKnown: process.env.HELSE_ID_WELL_KNOWN_URL,
    } satisfies Record<keyof ServerEnv, unknown | undefined>)
}

export const isLocalOrDemo = process.env.NODE_ENV === 'development' || bundledEnv.NEXT_PUBLIC_RUNTIME_ENV === 'demo'
export const isE2E = bundledEnv.NEXT_PUBLIC_RUNTIME_ENV === 'e2e'
