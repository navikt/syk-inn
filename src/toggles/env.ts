import { IToggle } from '@unleash/nextjs'

import { bundledEnv } from '@utils/env'

import { EXPECTED_TOGGLES } from './toggles'

export function localDevelopmentToggles(): IToggle[] {
    return EXPECTED_TOGGLES.map(
        (it): IToggle => ({
            name: it,
            enabled: false,
            impressionData: false,
            variant: {
                enabled: true,
                name: 'default',
            },
        }),
    )
}

export function getUnleashEnvironment(): 'development' | 'production' {
    if (bundledEnv.NEXT_PUBLIC_RUNTIME_ENV === 'prod-gcp') {
        return 'production'
    }
    return 'development'
}
