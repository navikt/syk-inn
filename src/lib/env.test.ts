import { beforeAll, beforeEach, describe, expect, test } from 'vitest'

import { getServerEnv } from './env'

describe('getServerEnv', () => {
    beforeAll(() => {
        process.env.HELSE_ID_WELL_KNOWN_URL = 'foo'
    })

    beforeEach(() => {
        delete process.env.VALKEY_HOST_SYK_INN
        delete process.env.VALKEY_USERNAME_SYK_INN
        delete process.env.VALKEY_PASSWORD_SYK_INN
    })

    describe('valkeyConfig', () => {
        test('should parse valkeyConfig for local', () => {
            process.env.VALKEY_HOST_SYK_INN = 'foo'

            const env = getServerEnv()

            expect(env.valkeyConfig).toEqual({
                runtimeEnv: 'local',
                host: 'foo',
            })
        })

        test('should parse valkeyConfig for dev', () => {
            process.env.NEXT_PUBLIC_RUNTIME_ENV = 'dev-gcp'
            process.env.VALKEY_HOST_SYK_INN = 'foo'
            process.env.VALKEY_USERNAME_SYK_INN = 'bar'
            process.env.VALKEY_PASSWORD_SYK_INN = 'baz'
            process.env.VALKEY_PORT_SYK_INN = '1234'

            const env = getServerEnv()

            expect(env.valkeyConfig).toEqual({
                runtimeEnv: 'dev-gcp',
                tls: {
                    host: 'foo',
                    port: 1234,
                },
                username: 'bar',
                password: 'baz',
            })
        })

        test('should parse valkeyConfig for prod', () => {
            process.env.NEXT_PUBLIC_RUNTIME_ENV = 'prod-gcp'
            process.env.VALKEY_HOST_SYK_INN = 'foo'
            process.env.VALKEY_USERNAME_SYK_INN = 'bar'
            process.env.VALKEY_PASSWORD_SYK_INN = 'baz'
            process.env.VALKEY_PORT_SYK_INN = '1234'

            const env = getServerEnv()

            expect(env.valkeyConfig).toEqual({
                runtimeEnv: 'prod-gcp',
                tls: {
                    host: 'foo',
                    port: 1234,
                },
                username: 'bar',
                password: 'baz',
            })
        })

        test('should parse valkeyConfig and throw if prod and missing username/password', () => {
            process.env.NEXT_PUBLIC_RUNTIME_ENV = 'prod-gcp'
            process.env.VALKEY_HOST_SYK_INN = 'foo'
            process.env.VALKEY_PORT_SYK_INN = '1234'

            expect(() => getServerEnv()).toThrowErrorMatchingSnapshot()
        })
    })
})
