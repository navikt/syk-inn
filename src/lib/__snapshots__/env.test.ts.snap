// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getServerEnv > valkeyConfig > should parse valkeyConfig and throw if prod and missing username/password 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "errors": [
      [
        {
          "expected": "string",
          "code": "invalid_type",
          "path": [
            "username"
          ],
          "message": "Invalid input: expected string, received undefined"
        },
        {
          "expected": "string",
          "code": "invalid_type",
          "path": [
            "password"
          ],
          "message": "Invalid input: expected string, received undefined"
        }
      ],
      [
        {
          "code": "invalid_value",
          "values": [
            "local"
          ],
          "path": [
            "runtimeEnv"
          ],
          "message": "Invalid input: expected \\"local\\""
        }
      ]
    ],
    "path": [
      "valkeyConfig"
    ],
    "message": "Invalid input"
  }
]]
`;
